File: DL-ATM41G2/DL-ATM41G2.php
<?php
/* https://www.decentlab.com/products/eleven-parameter-weather-station-for-lorawan */

abstract class DecentlabDecoder
{
    const PROTOCOL_VERSION = 2;

    protected $sensors;

    public function decode($payload = '')
    {
        $bytes = hex2bin($payload);
        $parts = [];

        $parts['version'] = ord($bytes[0]);
        if ($parts['version'] != self::PROTOCOL_VERSION) {
            $parts['error'] = sprintf("protocol version %u doesn't match v2", $parts['version']);
            return $parts;
        }

        $parts['device_id'] = unpack('n', substr($bytes, 1))[1];
        $flags = unpack('n', substr($bytes, 3))[1];

        /* decode payload */
        $k = 5;
        foreach ($this->sensors as $sensor) {
            if (($flags & 1) == 1) {
                $x = [];
                /* convert data to 16-bit integer array */
                for ($j = 0; $j < $sensor['length']; $j++) {
                    array_push($x, unpack('n', substr($bytes, $k))[1]);
                    $k += 2;
                }

                /* decode sensor values */
                foreach ($sensor['values'] as $value) {
                    if ($value['convert'] != NULL) {
                        $parts[$value['name'] . '_value'] = $value['convert']($x);
                        if ($value['unit'] != NULL) {
                            $parts[$value['name'] . '_unit'] = $value['unit'];
                        }
                    }
                }
            }
            $flags >>= 1;
        }

        return $parts;
    }
}

class DL_ATM41G2_Decoder extends DecentlabDecoder {
    
    public function __construct()
    {
        $this->sensors = [
            [
                'length' => 17,
                'values' => [
                    [
                        'name' => 'solar_radiation',
                        'convert' => function ($x) { return ($x[0] - 32768) / 10; },
                        'unit' => 'W⋅m⁻²',
                    ],
                    [
                        'name' => 'precipitation',
                        'convert' => function ($x) { return $x[1] / 1000; },
                        'unit' => 'mm',
                    ],
                    [
                        'name' => 'lightning_strike_count',
                        'convert' => function ($x) { return $x[2] - 32768; },
                        'unit' => 'None',
                    ],
                    [
                        'name' => 'lightning_average_distance',
                        'convert' => function ($x) { return $x[3] - 32768; },
                        'unit' => 'km',
                    ],
                    [
                        'name' => 'wind_speed',
                        'convert' => function ($x) { return ($x[4] - 32768) / 100; },
                        'unit' => 'm⋅s⁻¹',
                    ],
                    [
                        'name' => 'wind_direction',
                        'convert' => function ($x) { return ($x[5] - 32768) / 10; },
                        'unit' => '°',
                    ],
                    [
                        'name' => 'maximum_wind_speed',
                        'convert' => function ($x) { return ($x[6] - 32768) / 100; },
                        'unit' => 'm⋅s⁻¹',
                    ],
                    [
                        'name' => 'air_temperature',
                        'convert' => function ($x) { return ($x[7] - 32768) / 10; },
                        'unit' => '°C',
                    ],
                    [
                        'name' => 'vapor_pressure',
                        'convert' => function ($x) { return ($x[8] - 32768) / 100; },
                        'unit' => 'kPa',
                    ],
                    [
                        'name' => 'barometric_pressure',
                        'convert' => function ($x) { return ($x[9] - 32768) / 100; },
                        'unit' => 'kPa',
                    ],
                    [
                        'name' => 'relative_humidity',
                        'convert' => function ($x) { return ($x[10] - 32768) / 10; },
                        'unit' => '%',
                    ],
                    [
                        'name' => 'internal_temperature',
                        'convert' => function ($x) { return ($x[11] - 32768) / 10; },
                        'unit' => '°C',
                    ],
                    [
                        'name' => 'tilt_angle_x_orientation',
                        'convert' => function ($x) { return ($x[12] - 32768) / 10; },
                        'unit' => '°',
                    ],
                    [
                        'name' => 'tilt_angle_y_orientation',
                        'convert' => function ($x) { return ($x[13] - 32768) / 10; },
                        'unit' => '°',
                    ],
                    [
                        'name' => 'precipitation_electrical_conductivity',
                        'convert' => function ($x) { return $x[14] - 32768; },
                        'unit' => 'µS⋅cm⁻¹',
                    ],
                    [
                        'name' => 'cumulative_precipitation',
                        'convert' => function ($x) { return ($x[15] + $x[16] * 65536) / 1000; },
                        'unit' => 'mm',
                    ],
                ],
            ],
            [
                'length' => 1,
                'values' => [
                    [
                        'name' => 'battery_voltage',
                        'convert' => function ($x) { return $x[0] / 1000; },
                        'unit' => 'V',
                    ],
                ],
            ],
        ];
    }
}


$decoder = new DL_ATM41G2_Decoder();
$payloads = [
    '025ef80003805c000080008000803484b3803680e78086a60181d680ed81c9809f8000117000010adc',
    '025ef800020adc',
];

foreach($payloads as $payload) {
    var_dump($decoder->decode($payload));
}

File: DL-ATM41G2/DL-ATM41G2.py
# -*- coding: utf-8 -*-

# https://www.decentlab.com/products/eleven-parameter-weather-station-for-lorawan

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import struct
from math import log, floor
from base64 import binascii

PROTOCOL_VERSION = 2

SENSORS = [
    {'length': 17,
     'values': [{'name': 'Solar radiation',
                 'convert': lambda x: (x[0] - 32768) / 10,
                 'unit': 'W⋅m⁻²'},
                {'name': 'Precipitation',
                 'convert': lambda x: x[1] / 1000,
                 'unit': 'mm'},
                {'name': 'Lightning strike count',
                 'convert': lambda x: x[2] - 32768,
                 'unit': 'None'},
                {'name': 'Lightning average distance',
                 'convert': lambda x: x[3] - 32768,
                 'unit': 'km'},
                {'name': 'Wind speed',
                 'convert': lambda x: (x[4] - 32768) / 100,
                 'unit': 'm⋅s⁻¹'},
                {'name': 'Wind direction',
                 'convert': lambda x: (x[5] - 32768) / 10,
                 'unit': '°'},
                {'name': 'Maximum wind speed',
                 'convert': lambda x: (x[6] - 32768) / 100,
                 'unit': 'm⋅s⁻¹'},
                {'name': 'Air temperature',
                 'convert': lambda x: (x[7] - 32768) / 10,
                 'unit': '°C'},
                {'name': 'Vapor pressure',
                 'convert': lambda x: (x[8] - 32768) / 100,
                 'unit': 'kPa'},
                {'name': 'Barometric pressure',
                 'convert': lambda x: (x[9] - 32768) / 100,
                 'unit': 'kPa'},
                {'name': 'Relative humidity',
                 'convert': lambda x: (x[10] - 32768) / 10,
                 'unit': '%'},
                {'name': 'Internal temperature',
                 'convert': lambda x: (x[11] - 32768) / 10,
                 'unit': '°C'},
                {'name': 'Tilt angle, X orientation',
                 'convert': lambda x: (x[12] - 32768) / 10,
                 'unit': '°'},
                {'name': 'Tilt angle, Y orientation',
                 'convert': lambda x: (x[13] - 32768) / 10,
                 'unit': '°'},
                {'name': 'Precipitation electrical conductivity',
                 'convert': lambda x: x[14] - 32768,
                 'unit': 'µS⋅cm⁻¹'},
                {'name': 'Cumulative precipitation',
                 'convert': lambda x: (x[15] + x[16] * 65536) / 1000,
                 'unit': 'mm'}]},
    {'length': 1,
     'values': [{'name': 'Battery voltage',
                 'convert': lambda x: x[0] / 1000,
                 'unit': 'V'}]}
]


def decode(msg, hex=False):
    """msg: payload as one of hex string, list, or bytearray"""
    bytes_ = bytearray(binascii.a2b_hex(msg)
                       if hex
                       else msg)
    version = bytes_[0]
    if version != PROTOCOL_VERSION:
        raise ValueError("protocol version {} doesn't match v2".format(version))

    devid = struct.unpack('>H', bytes_[1:3])[0]
    bin_flags = bin(struct.unpack('>H', bytes_[3:5])[0])
    flags = bin_flags[2:].zfill(struct.calcsize('>H') * 8)[::-1]

    words = [struct.unpack('>H', bytes_[i:i + 2])[0]
             for i
             in range(5, len(bytes_), 2)]

    cur = 0
    result = {'Device ID': devid, 'Protocol version': version}
    for flag, sensor in zip(flags, SENSORS):
        if flag != '1':
            continue

        x = words[cur:cur + sensor["length"]]
        cur += sensor["length"]
        for value in sensor['values']:
            if 'convert' not in value:
                continue

            result[value['name']] = {'value': value['convert'](x),
                                     'unit': value.get('unit', None)}

    return result


if __name__ == '__main__':

    import pprint
    payloads = [
        b'025ef80003805c000080008000803484b3803680e78086a60181d680ed81c9809f8000117000010adc',
        b'025ef800020adc',
    ]
    for pl in payloads:
        pprint.pprint(decode(pl, hex=True))
        print("")
